packages/clerk-js/src/ui/components/SignUp/signUpFormHelpers.ts

```diff
@@ -300,3 +304,124 @@ function getGenericField(fieldKey: FieldKey, attributes: Partial<Attributes>): F
+  username: SignUpAttributeField;
+};
+
+const outcomePredicates: Record<Outcome, ((ctx: Context) => boolean)[]> = {
+  mirrorServer: [
+    // If password is not required, then field requirements are determined by the server.
+    ctx => !ctx.passwordRequired,
+    // If any of the communication methods are already required by the server, then we don't need to do anything.
+    ctx => ctx.email.required || ctx.phone.required || (ctx.username.required && ctx.username.firstFactor),
+  ],
+  none: [
+    // If none of the communication methods are enabled, then none can be required.
+    ctx => !ctx.email.enabled && !ctx.phone.enabled && !ctx.username.enabled,
+  ],
+  email: [
+    // If email is the only enabled communication method, it should be required.
+    ctx => ctx.email.enabled && !ctx.phone.enabled && !ctx.username.enabled,
+    // If email is enabled but not required, and phone is enabled and not required, then email should be required.
+    ctx => ctx.email.enabled && !ctx.email.required && ctx.phone.enabled && !ctx.phone.required,
+    // If username is a first factor but not required, email can be used as an alternative.
+    ctx => ctx.username.firstFactor && !ctx.username.required && ctx.email.enabled && !ctx.email.required,
+    // If username is required but not a first factor, and both email and phone are enabled, then email is a valid identifier.
+    ctx => ctx.username.required && !ctx.username.firstFactor && ctx.email.enabled && ctx.phone.enabled,
+  ],
+  phone: [
+    ctx => ctx.phone.enabled && !ctx.email.required && !ctx.phone.required,
+    // If username is a first factor but not required, phone can be used as an alternative.
+    ctx => ctx.username.firstFactor && !ctx.username.required && ctx.phone.enabled && !ctx.phone.required,
+    // If phone is the only first factor, it should be required.
+    ctx => ctx.phone.firstFactor && !ctx.email.firstFactor && !ctx.username.firstFactor,
+    // If username is required but not a first factor, and both email and phone are enabled, then phone is a valid identifier.
+    ctx => ctx.username.required && !ctx.username.firstFactor && ctx.phone.enabled && ctx.email.enabled,
+    // If email is not enabled, but phone and username are, phone should be available.
+    ctx => !ctx.email.enabled && ctx.phone.enabled && ctx.username.enabled,
+  ],
+  username: [
+    // If username is the only first factor, it should be required.
+    ctx => ctx.username.enabled && ctx.username.firstFactor && !ctx.email.enabled && !ctx.phone.enabled,
+    // If username is required but not a first factor, and both email and phone are enabled, it should be required.
+    ctx => ctx.username.required && !ctx.username.firstFactor && ctx.email.enabled && ctx.phone.enabled,
+  ],
+};
+
+/**
+ * When password is required, we need to ensure at least one identifier
+ * (email, phone, or username) is also required
+ */
+export function determineRequiredCommunicationMethod(attributes: Partial<Attributes>): {
```

Comment:
nit: for consistency, I might consider naming this `determineRequiredIdentifier`. This avoids introducing another term "communication method" to describe these things.